File: frontend-react/src/api/mockProvider.ts
import type { Candidate, PromotionState } from "./candidateModel";
import { candidateFromMapping } from "./candidateModel";

let store: Candidate[] = [];

export async function loadMockData(): Promise<void> {
  const res = await fetch("/mock-candidates.json");
  const raw = (await res.json()) as Record<string, unknown>[];
  store = raw.map(candidateFromMapping);
}

export function listCandidates(state?: PromotionState): Candidate[] {
  if (!state) return [...store];
  return store.filter((c) => c.state === state);
}

export function getCandidate(id: string): Candidate | undefined {
  return store.find((c) => c.id === id);
}

export function promote(id: string, targetState?: PromotionState): Candidate {
  const idx = store.findIndex((c) => c.id === id);
  if (idx === -1) throw new Error("not found");
  const cand = store[idx];
  const next = targetState ?? nextPromotionState(cand.state as PromotionState);
  if (!next) throw new Error("terminal");
  const updated: Candidate = {
    ...cand,
    state: next,
    auditTrail: [
      ...(cand.auditTrail ?? []),
      {
        action: "promote",
        timestamp: new Date().toISOString(),
        provenance: "react-mock",
        actor: "human",
        note: `to ${next}`,
      },
    ],
  };
  store[idx] = updated;
  return updated;
}

export function reject(id: string, reason?: string): Candidate {
  const idx = store.findIndex((c) => c.id === id);
  if (idx === -1) throw new Error("not found");
  const updated: Candidate = {
    ...store[idx],
    state: "decayed",
    auditTrail: [
      ...(store[idx].auditTrail ?? []),
      {
        action: "reject",
        timestamp: new Date().toISOString(),
        provenance: "react-mock",
        actor: "human",
        note: reason,
      },
    ],
  };
  store[idx] = updated;
  return updated;
}

export function resolveContradiction(
  primaryId: string,
  rivalId: string,
  resolution: "keep_a" | "keep_b",
  keepId: string
): Candidate {
  const keptIdx = store.findIndex((c) => c.id === keepId);
  if (keptIdx === -1) throw new Error("keep target not found");
  const loserId = keepId === primaryId ? rivalId : primaryId;
  const loserIdx = store.findIndex((c) => c.id === loserId);
  if (loserIdx !== -1) {
    store[loserIdx] = {
      ...store[loserIdx],
      state: "decayed",
      auditTrail: [
        ...(store[loserIdx].auditTrail ?? []),
        {
          action: "resolve_contradiction_loser",
          timestamp: new Date().toISOString(),
          provenance: "react-mock",
          actor: "human",
          note: `lost to ${keepId}`,
        },
      ],
    };
  }
  const kept = store[keptIdx];
  store[keptIdx] = {
    ...kept,
    auditTrail: [
      ...(kept.auditTrail ?? []),
      {
        action: "resolve_contradiction",
        timestamp: new Date().toISOString(),
        provenance: "react-mock",
        actor: "human",
        note: `${resolution} kept ${keepId}`,
      },
    ],
  };
  return store[keptIdx];
}

function nextPromotionState(state: PromotionState): PromotionState | undefined {
  if (state === "proposed") return "suggested";
  if (state === "suggested") return "active";
  return undefined;
}

File: frontend/services/mock_provider.py
"""In-memory mock provider for offline demo and contract tests."""

from __future__ import annotations

from copy import deepcopy
from datetime import datetime, timezone
from typing import Any

from ..models.candidate import Candidate, PromotionState
from .data_provider import DataProvider


class MockDataProvider(DataProvider):
    """Fully functional in-memory implementation.

    Used when PRAXIS_API_BASE_URL is unset. Supports the full promote/reject/resolve
    lifecycle and maintains an audit trail on each mutation.
    """

    def __init__(self, seed: list[dict[str, Any]] | None = None) -> None:
        from ..mock_data import SEED_CANDIDATES

        raw = seed or SEED_CANDIDATES
        self._store: dict[str, Candidate] = {c.id: deepcopy(c) for c in raw}
        self._contradiction_pairs: set[tuple[str, str]] = set()

    # ------------------------------------------------------------------ queries
    def list_candidates(self, state: PromotionState | None = None) -> list[Candidate]:
        if state is None:
            return list(self._store.values())
        return [c for c in self._store.values() if c.state == state]

    def get_candidate(self, candidate_id: str) -> Candidate | None:
        return self._store.get(candidate_id)

    # ------------------------------------------------------------------ mutations
    def promote(self, candidate_id: str, target_state: PromotionState | None = None) -> Candidate:
        cand = self._require(candidate_id)
        next_state = target_state or cand.next_promotion_state()
        if next_state is None:
            raise ValueError(f"Cannot promote terminal state {cand.state}")
        self._append_audit(cand, "promote", note=f"to {next_state}")
        cand.state = next_state
        return cand

    def reject(self, candidate_id: str, reason: str | None = None) -> Candidate:
        cand = self._require(candidate_id)
        self._append_audit(cand, "reject", note=reason)
        cand.state = "decayed"
        return cand

    def resolve_contradiction(
        self, primary_id: str, rival_id: str, resolution: str, keep_id: str
    ) -> Candidate:
        kept = self._require(keep_id)
        self._append_audit(
            kept,
            "resolve_contradiction",
            note=f"{resolution} (kept {keep_id}) vs {primary_id}/{rival_id}",
        )
        # Mark the loser as decayed for demo purposes
        loser_id = primary_id if keep_id == rival_id else rival_id
        if loser_id in self._store:
            loser = self._store[loser_id]
            loser.state = "decayed"
            self._append_audit(loser, "resolve_contradiction_loser", note=f"lost to {keep_id}")
        self._contradiction_pairs.add(tuple(sorted([primary_id, rival_id])))
        return kept

    # ------------------------------------------------------------------ helpers
    def _require(self, candidate_id: str) -> Candidate:
        if candidate_id not in self._store:
            raise KeyError(candidate_id)
        return self._store[candidate_id]

    def _append_audit(self, cand: Candidate, action: str, note: str | None = None) -> None:
        entry: dict[str, Any] = {
            "action": action,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "provenance": "mock_provider",
            "actor": "human",
        }
        if note:
            entry["note"] = note
        if cand.audit_trail is None:
            cand.audit_trail = []
        cand.audit_trail.append(entry)
